### PR TITLE
fix(levm): change gas type from U256 to u64

### PR DESCRIPTION
**Pull request by lima-limon-inc**
**Motivation**

The execution client uses `u64` for measuring gas, while LEVM uses `U256`. This PR makes LEVM also use u64 for gas 

**Description**

Closes https://github.com/lambdaclass/ethrex/issues/1499

### CODE REVIEW

**Comment by lima-limon-inc** [comment]
We'll see when it finishes running, but in theory all the uses of gas have been changed to u64. 

The files that relate to parsing the tests and running the benchmarks remain in U256, simply to maintain how the parsing works. Those values are cast to u64 when the tests run.

**Comment by lima-limon-inc** [comment]
~PS: Note to reviewers~

~I used the "VeryLargeNumber" when a function couldn't cast a value from U256 `u64`. Should I return a different error?
What error should it be?~

Changed here: https://github.com/lambdaclass/ethrex/pull/1528/commits/7b3bccc892b93dcfaadfed8a40be3e2e0d18c511

**Review by damiramirez — APPROVED** [review]
LGTM

File: crates/vm/levm/src/gas_cost.rs [hunk 1]
```diff
@@ -180,27 +180,31 @@ pub const MODEXP_STATIC_COST: u64 = 0;
 pub const MODEXP_DYNAMIC_BASE: u64 = 200;
 pub const MODEXP_DYNAMIC_QUOTIENT: u64 = 3;
 
-pub fn exp(exponent: U256) -> Result<U256, OutOfGasError> {
+pub fn exp(exponent: U256) -> Result<u64, VMError> {
     let exponent_byte_size = (exponent
         .bits()
         .checked_add(7)
-        .ok_or(OutOfGasError::GasCostOverflow)?)
+        .ok_or(VMError::OutOfGas(OutOfGasError::GasCostOverflow))?)
```
Comment (JereSalo) [inline crates/vm/levm/src/gas_cost.rs:187]: It wasn't neceesary to add `VMError::OutOfGas`. `thiserror` crate handles that. Doesn't matter much though, just a thing i noticed

File: crates/vm/levm/src/gas_cost.rs [hunk 2]
```diff
@@ -180,27 +180,31 @@ pub const MODEXP_STATIC_COST: u64 = 0;
 pub const MODEXP_DYNAMIC_BASE: u64 = 200;
 pub const MODEXP_DYNAMIC_QUOTIENT: u64 = 3;
 
-pub fn exp(exponent: U256) -> Result<U256, OutOfGasError> {
+pub fn exp(exponent: U256) -> Result<u64, VMError> {
```
Comment (JereSalo) [inline crates/vm/levm/src/gas_cost.rs:183]: `OutOfGasError` was a good fit for this function, it says explicitly that it can fail only because of that. Doesn't matter much though.

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 3]
```diff
@@ -23,7 +23,11 @@ impl VM {
         current_call_frame: &mut CallFrame,
     ) -> Result<OpcodeSuccess, VMError> {
         // STACK
-        let gas = current_call_frame.stack.pop()?;
+        let gas: u64 = current_call_frame
+            .stack
+            .pop()?
+            .try_into()
+            .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/system.rs:30]: This shouldn't be internal

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 4]
```diff
@@ -99,7 +103,11 @@ impl VM {
...
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/system.rs:110]: ditto

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 5]
```diff
@@ -200,7 +211,11 @@ impl VM {
         current_call_frame: &mut CallFrame,
     ) -> Result<OpcodeSuccess, VMError> {
         // STACK
-        let gas = current_call_frame.stack.pop()?;
+        let gas = current_call_frame
+            .stack
+            .pop()?
+            .try_into()
+            .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/system.rs:218]: ditto

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 6]
```diff
@@ -258,7 +273,11 @@ impl VM {
...
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/system.rs:280]: ditto

**Review by JereSalo — CHANGES_REQUESTED** [review]
Great job! There was a lot of effort in this

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 7]
```diff
@@ -180,10 +188,13 @@ impl VM {
...
         let new_memory_size = calculate_memory_size(offset, size)?;
-        self.increase_consumed_gas(
-            current_call_frame,
-            memory::expansion_cost(new_memory_size, current_call_frame.memory.len())?.into(),
-        )?;
+
+        let memory_expansion_cost: u64 =
+            memory::expansion_cost(new_memory_size, current_call_frame.memory.len())?
+                .try_into()
+                .map_err(|_err| VMError::Internal(InternalError::ConversionError))?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/opcode_handlers/system.rs:195]: Should this conversion error remain?

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 8]
```diff
@@ -396,10 +415,13 @@ impl VM {
...
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/opcode_handlers/system.rs:422]: Ditto

File: crates/vm/levm/src/vm.rs [hunk 9]
```diff
@@ -428,11 +428,15 @@ impl VM {
...
                         let code_length = contract_code.len();
-                        let code_deposit_cost = U256::from(code_length)
-                            .checked_mul(CODE_DEPOSIT_COST)
-                            .ok_or(VMError::Internal(
-                                InternalError::ArithmeticOperationOverflow,
-                            ))?;
+
+                        let code_length_u64: u64 = code_length
+                            .try_into()
+                            .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/vm.rs:434]: Ditto

File: crates/vm/levm/src/gas_cost.rs (hunk 2 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/gas_cost.rs:183]: Hi Jere, good morning. Now I remember why I changed the error type. 
It's because of this casting that's  below (The one that returns "VeryLargeNumber")

File: crates/vm/levm/src/gas_cost.rs [hunk 10]
```diff
@@ -282,48 +293,59 @@ pub fn log(
     current_memory_size: usize,
     size: usize,
     number_of_topics: u8,
-) -> Result<U256, VMError> {
+) -> Result<u64, VMError> {
     let memory_expansion_cost = memory::expansion_cost(new_memory_size, current_memory_size)?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/gas_cost.rs:297]: See: https://github.com/lambdaclass/ethrex/issues/1547

File: cmd/ef_tests/levm/runner/revm_runner.rs [hunk 11]
```diff
@@ -101,11 +101,13 @@ pub fn prepare_revm_for_tx<'state>(
...
         .chain_config()
         .map_err(|err| EFTestRunnerError::VMInitializationFailed(err.to_string()))?;
+
+    let gas_limit: U256 = test.env.current_gas_limit.into();
     let block_env = RevmBlockEnv {
         number: RevmU256::from_limbs(test.env.current_number.0),
         coinbase: RevmAddress(test.env.current_coinbase.0.into()),
         timestamp: RevmU256::from_limbs(test.env.current_timestamp.0),
-        gas_limit: RevmU256::from_limbs(test.env.current_gas_limit.0),
+        gas_limit: RevmU256::from_limbs(gas_limit.0),
```
Comment (JereSalo) [inline cmd/ef_tests/levm/runner/revm_runner.rs:110]: Isn't there here a straightforward `::from(u64)` method? If so you can avoid having a `gas_limit` variable. Otherwise, I suggest renaming that to `block_gas_limit` to avoid confusion.

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 12]
```diff
@@ -23,7 +23,9 @@ impl VM {
         current_call_frame: &mut CallFrame,
     ) -> Result<OpcodeSuccess, VMError> {
         // STACK
-        let gas = current_call_frame.stack.pop()?;
+        let gas: U256 = current_call_frame.stack.pop()?;
+        // dbg!(gas);
+        // let gas: u64 = gas.try_into().map_err(|_| VMError::BalanceOverflow)?;
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/system.rs:28]: oopsie

File: crates/vm/levm/src/vm.rs [hunk 13]
```diff
@@ -558,18 +562,19 @@ impl VM {
                 .ok_or(OutOfGasError::ConsumedGasOverflow)?;
 
             let number_of_words = initial_call_frame.calldata.len().div_ceil(WORD_SIZE);
+            let double_nuber_of_words: u64 = number_of_words
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:565]:
```suggestion
            let double_number_of_words: u64 = number_of_words
```
Suggestion, in this kind of PRs that are very big maybe it is best not to do these kind of changes, but I agree that it looks better to have this in another variable.

File: crates/vm/levm/src/vm.rs [hunk 14]
```diff
@@ -592,37 +597,50 @@ impl VM {
...
 
     /// Gets the max blob gas cost for a transaction that a user is willing to pay.
-    fn get_max_blob_gas_cost(&self) -> Result<U256, VMError> {
-        let blob_gas_used = U256::from(self.env.tx_blob_hashes.len())
+    fn get_max_blob_gas_price(&self) -> Result<U256, VMError> {
+        let blob_gasses = self.env.tx_blob_hashes.len();
+
+        let blob_gas_used: u64 = blob_gasses
+            .try_into()
+            .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:605]: Can we change these names being used? Just leaving only one `let blob_gas_used` will be fine. If you want to can rename the blob hashes len to `blob_hashes_len` or something similar, but not `blob_gasses`

File: crates/vm/levm/src/vm.rs [hunk 15]
```diff
@@ -592,37 +597,50 @@ impl VM {
...
             .checked_mul(BLOB_GAS_PER_BLOB)
             .unwrap_or_default();
 
         let base_fee_per_blob_gas = self.get_base_fee_per_blob_gas()?;
 
-        let blob_fee = blob_gas_used
+        let blob_fee = blob_gas_price
             .checked_mul(base_fee_per_blob_gas)
-            .ok_or(InternalError::UndefinedState(1))?;
+            .ok_or(VMError::Internal(InternalError::ConversionError))?;
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:638]: Again, if possible don't change things that are not directly related to the PR's objective, specially if it is a huge refactor PR

**Review by JereSalo — CHANGES_REQUESTED** [review]
I left some suggestions. If possible re-review the PR yourself and see if you find anything out of place. I think other than the things I mentioned it looks good

File: crates/vm/levm/src/opcode_handlers/system.rs (hunk 12 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/opcode_handlers/system.rs]: https://github.com/lambdaclass/ethrex/pull/1528/commits/52198065ca30affc698038209cf09070788348c2

File: crates/vm/levm/src/vm.rs (hunk 13 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs:565]: Noted Jere, I only added a new variable in the cases where a variable was overloaded with method calls after I added the new types. 

Should I keep _this_ specific variable? I'll change the rest

Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Spellcheck: https://github.com/lambdaclass/ethrex/pull/1528/commits/bab45f27990b7a9ef3c86cb6c5b33da442ec4dcc

File: crates/vm/levm/src/vm.rs (hunk 14 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Changed here: https://github.com/lambdaclass/ethrex/pull/1528/commits/1e5bf00c8712e96da3ca53bff5ebfdcfb0879dcf

File: crates/vm/levm/src/vm.rs (hunk 15 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Changed here: 
https://github.com/lambdaclass/ethrex/pull/1528/commits/17e8c68cc07b836fbdb220b0d25695bee237c1e5

**Comment by lima-limon-inc** [comment]
> I left some suggestions. If possible re-review the PR yourself and see if you find anything out of place. I think other than the things I mentioned it looks good

Hello Jere, thank you very much for your feedback. I incorporated your suggestions and re-reviewed the PR. Let me know if you find anything else that need correction.

PS: I also found some other places (https://github.com/lambdaclass/ethrex/pull/1528/commits/ef21d78038065fcd1b38b2a31146cc6b7b7c953d) where I had created a variable where there wasn't any before. I rolled back those changes so that is more _aligned_ with the previous style. 

LMKWYT!

File: crates/vm/levm/src/vm.rs [hunk 16]
```diff
@@ -610,37 +615,51 @@ impl VM {
     }
 
     /// Gets the max blob gas cost for a transaction that a user is willing to pay.
-    fn get_max_blob_gas_cost(&self) -> Result<U256, VMError> {
-        let blob_gas_used = U256::from(self.env.tx_blob_hashes.len())
+    fn get_max_blob_gas_price(&self) -> Result<U256, VMError> {
+        let blob_gas_used: u64 = self
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:619]: I know you rewrite this but this is not `blob_gas_used`

File: crates/vm/levm/src/vm.rs [hunk 17]
```diff
@@ -610,37 +615,51 @@ impl VM {
...
             .ok_or(InternalError::UndefinedState(1))?;
 
         Ok(max_blob_gas_cost)
     }
 
     /// Gets the actual blob gas cost.
-    fn get_blob_gas_cost(&self) -> Result<U256, VMError> {
-        let blob_gas_used = U256::from(self.env.tx_blob_hashes.len())
+    fn get_blob_gas_price(&self) -> Result<U256, VMError> {
+        let blob_gas_used: u64 = self
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:642]: ditto

File: crates/vm/levm/src/vm.rs [hunk 18]
```diff
@@ -610,37 +615,51 @@ impl VM {
...
             .checked_mul(base_fee_per_blob_gas)
-            .ok_or(InternalError::UndefinedState(1))?;
+            .ok_or(VMError::Internal(InternalError::UndefinedState(1)))?;
 
-        Ok(blob_fee)
+        Ok(blob_fee.into())
     }
 
-    pub fn get_base_fee_per_blob_gas(&self) -> Result<U256, VMError> {
+    pub fn get_base_fee_per_blob_gas(&self) -> Result<u64, VMError> {
```
Comment (JereSalo) [inline crates/vm/levm/src/vm.rs:662]: this sounds like it should be an U256 because it is a fee. I am not 100% sure but I think it should because in the validations you are comparing it against `tx_max_fee_per_blob_gas`

File: crates/vm/levm/src/vm.rs (hunk 16 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Noted, I changed it to `blob_gas_amout` (https://github.com/lambdaclass/ethrex/pull/1528/commits/e0a5eec01c68aa16348e9deb7743e31c16e58498), what do you think?

File: crates/vm/levm/src/vm.rs (hunk 17 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: ditto ditto: https://github.com/lambdaclass/ethrex/pull/1528/commits/e235e99b2c001ffbd45a4442cd4fd67117c51882

File: crates/vm/levm/src/vm.rs (hunk 18 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Hello Jere, this was changed here: https://github.com/lambdaclass/ethrex/pull/1528/commits/96b9bf52d81d92cbc91ebccd277e7eb33b2a3375

I added an additional let binding  (`let blob_gas_price: U256 = blob_gas_price.into();`, https://github.com/lambdaclass/ethrex/pull/1528/commits/96b9bf52d81d92cbc91ebccd277e7eb33b2a3375#diff-e20d30a6efb08393eee586d44f3bc4a8df6cbd53ec8ea5bb4d5def88c9487de1R655) simply out of necessity. It was the cleanest way I could think of to cast the value to U256.

File: crates/vm/levm/src/vm.rs (hunk 16 above)
Comment (JereSalo) [reply crates/vm/levm/src/vm.rs]: Well, it is actually not `blob_gas_amount`, in that case it is just the length of the blob hash list! Not related to gas

Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs]: Sorry Jere, it was reading the variable name wrong in my head. It's been a rough day! Haha,

Changed here: https://github.com/lambdaclass/ethrex/pull/1528/commits/341ca8d2ccf93ab7e932dc7fa8b8fdc05d405cf4

**Review by JereSalo — APPROVED** [review]
nice! sorry if I was pretty annoying with the suggestions. good work

**Comment by lima-limon-inc** [comment]
> nice! sorry if I was pretty annoying with the suggestions. good work

Please Jere, no need to apologize! On the contrary, I not only value your diligence, I also admire it!

I have never taken any offense from your comments. Thank you very much for all your dedication! 

:blush:
